The build error suggests there need package imports of org.apache.batik.transcoder and org.apache.batik.transcoder.print  in MANIFEST.MF of the test bundle where you added the test.

```
09:37:44  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:5.0.2:compile (default-compile) on project org.eclipse.birt.report.engine.tests: Compilation failure: Compilation failure: 
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[7] 
09:37:44  [ERROR] 	import org.apache.batik.transcoder.SVGAbstractTranscoder;
09:37:44  [ERROR] 	       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
09:37:44  [ERROR] The import org.apache.batik.transcoder cannot be resolved
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[8] 
09:37:44  [ERROR] 	import org.apache.batik.transcoder.TranscoderInput;
09:37:44  [ERROR] 	       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
09:37:44  [ERROR] The import org.apache.batik.transcoder cannot be resolved
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[9] 
09:37:44  [ERROR] 	import org.apache.batik.transcoder.print.PrintTranscoder;
09:37:44  [ERROR] 	       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
09:37:44  [ERROR] The import org.apache.batik.transcoder cannot be resolved
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[159] 
09:37:44  [ERROR] 	PrintTranscoder transcoder = new PrintTranscoder();
09:37:44  [ERROR] 	^^^^^^^^^^^^^^^
09:37:44  [ERROR] PrintTranscoder cannot be resolved to a type
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[159] 
09:37:44  [ERROR] 	PrintTranscoder transcoder = new PrintTranscoder();
09:37:44  [ERROR] 	                                 ^^^^^^^^^^^^^^^
09:37:44  [ERROR] PrintTranscoder cannot be resolved to a type
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[161] 
09:37:44  [ERROR] 	transcoder.transcode(new TranscoderInput(new ByteArrayInputStream(svgData)), null);
09:37:44  [ERROR] 	                         ^^^^^^^^^^^^^^^
09:37:44  [ERROR] TranscoderInput cannot be resolved to a type
09:37:44  [ERROR] /home/jenkins/agent/workspace/build_PR-2409/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java:[163] 
09:37:44  [ERROR] 	transcoder.addTranscodingHint(SVGAbstractTranscoder.KEY_ALLOW_EXTERNAL_RESOURCES, Boolean.TRUE);
09:37:44  [ERROR] 	                              ^^^^^^^^^^^^^^^^^^^^^
09:37:44  [ERROR] SVGAbstractTranscoder cannot be resolved to a variable
09:37:44  [ERROR] 7 problems (7 errors)
```

--- a/engine/org.eclipse.birt.report.engine.emitter.pdf/src/org/eclipse/birt/report/engine/emitter/pdf/PDFPage.java
+++ b/engine/org.eclipse.birt.report.engine.emitter.pdf/src/org/eclipse/birt/report/engine/emitter/pdf/PDFPage.java
@@ -185,9 +185,14 @@ public class PDFPage extends AbstractPage {
 
 			// SVG images
 			try {
-				if (this.pageDevice.useBackgroundImageSvg() && SvgFile.isSvg(null, null, imageUrl)
-						&& (new File((new URL(imageUrl)).toURI().getPath())).exists()) {
-					image = transSVG(imageUrl, null, imageHeight, imageWidth);
+				boolean isSvgImage = this.pageDevice.useBackgroundImageSvg()
+						&& SvgFile.isSvg(null, imageUrl, null, imageData);
+				if (isSvgImage) {
+					if ((new File((new URL(imageUrl)).toURI().getPath())).exists()) {
+						image = transSVG(imageUrl, null, imageHeight, imageWidth);
+					} else if (imageData != null) {
+						image = transSVG(null, imageData, imageHeight, imageWidth);
+					}
 					if (image != null) {
 						drawImage(image, x, y, imageHeight, imageWidth, null);
 					}
@@ -283,7 +288,7 @@ public class PDFPage extends AbstractPage {
 		}
 
 		// Not cached yet
-		if (SvgFile.isSvg(null, null, extension)) {
+		if (SvgFile.isSvg(null, null, extension, imageData)) {
 			template = generateTemplateFromSVG(imageData, height, width);
 		} else {
 			// PNG/JPG/BMP... images:

--- a/engine/org.eclipse.birt.report.engine.tests/META-INF/MANIFEST.MF
+++ b/engine/org.eclipse.birt.report.engine.tests/META-INF/MANIFEST.MF
@@ -53,3 +53,6 @@ Export-Package: org.eclipse.birt.report.engine;x-internal:=true,
  org.eclipse.birt.report.engine.tests;x-internal:=true,
  org.eclipse.birt.report.engine.toc;x-internal:=true,
  org.eclipse.birt.report.engine.util;x-internal:=true
+Import-Package: org.apache.batik.transcoder;version="1.19.0",
+ org.apache.batik.transcoder.print;version="1.19.0"
+

--- a/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java
+++ b/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/emitter/pdf/PDFSvgEmbeddingTest.java
@@ -1,0 +1,190 @@
+/*******************************************************************************
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0/.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   See git history
+ *******************************************************************************/
+
+package org.eclipse.birt.report.engine.emitter.pdf;
+
+import java.io.ByteArrayOutputStream;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.batik.transcoder.SVGAbstractTranscoder;
+import org.apache.batik.transcoder.TranscoderInput;
+import org.apache.batik.transcoder.print.PrintTranscoder;
+import org.openpdf.text.Document;
+import org.openpdf.text.PageSize;
+import org.openpdf.text.pdf.PdfContentByte;
+import org.openpdf.text.pdf.PdfTemplate;
+import org.openpdf.text.pdf.PdfWriter;
+
+import junit.framework.TestCase;
+
+public class PDFSvgEmbeddingTest extends TestCase {
+
+	private static final String SIMPLE_SVG =
+			"<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"100\" height=\"100\">" +
+			"<rect x=\"10\" y=\"10\" width=\"80\" height=\"80\" fill=\"red\"/>" +
+			"</svg>";
+
+	private static final String SVG_WITH_PATH =
+			"<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"200\" height=\"200\">" +
+			"<path d=\"M10,10 L190,10 L190,190 L10,190 Z\" fill=\"blue\" stroke=\"black\" stroke-width=\"2\"/>" +
+			"</svg>";
+
+	private static final String SVG_WITH_CIRCLE =
+			"<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"50\" height=\"50\">" +
+			"<circle cx=\"25\" cy=\"25\" r=\"20\" fill=\"green\"/>" +
+			"</svg>";
+
+	private static final String SVG_WITH_TEXT =
+			"<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"300\" height=\"50\">" +
+			"<text x=\"10\" y=\"30\" font-family=\"Arial\" font-size=\"20\">Hello World</text>" +
+			"</svg>";
+
+	public void testTranscodeSvgByteArrayToPdfTemplate() throws Exception {
+		byte[] svgData = SIMPLE_SVG.getBytes(StandardCharsets.UTF_8);
+		float width = 100, height = 100;
+
+		PdfTemplate template = transSvgToTemplate(svgData, width, height);
+
+		assertNotNull("PDF template must not be null", template);
+		assertTrue("Template width must be positive", template.getWidth() > 0);
+		assertTrue("Template height must be positive", template.getHeight() > 0);
+	}
+
+	public void testTranscodeSvgWithPath() throws Exception {
+		byte[] svgData = SVG_WITH_PATH.getBytes(StandardCharsets.UTF_8);
+		float width = 200, height = 200;
+
+		PdfTemplate template = transSvgToTemplate(svgData, width, height);
+
+		assertNotNull("PDF template must not be null", template);
+		assertEquals("Template width must match", width, template.getWidth());
+		assertEquals("Template height must match", height, template.getHeight());
+	}
+
+	public void testTranscodeSvgWithCircle() throws Exception {
+		byte[] svgData = SVG_WITH_CIRCLE.getBytes(StandardCharsets.UTF_8);
+		PdfTemplate template = transSvgToTemplate(svgData, 50, 50);
+		assertNotNull("PDF template for circle SVG must not be null", template);
+	}
+
+	public void testTranscodeSvgWithText() throws Exception {
+		byte[] svgData = SVG_WITH_TEXT.getBytes(StandardCharsets.UTF_8);
+		PdfTemplate template = transSvgToTemplate(svgData, 300, 50);
+		assertNotNull("PDF template for text SVG must not be null", template);
+	}
+
+	public void testGeneratedPdfIsReadableAndContainsSvgContent() throws Exception {
+		byte[] svgData = SVG_WITH_PATH.getBytes(StandardCharsets.UTF_8);
+
+		Document document = new Document(PageSize.A4);
+		ByteArrayOutputStream pdfOut = new ByteArrayOutputStream();
+		PdfWriter writer = PdfWriter.getInstance(document, pdfOut);
+
+		document.open();
+
+		PdfContentByte cb = writer.getDirectContent();
+		PdfTemplate svgTemplate = transSvgToTemplate(svgData, 200, 200);
+
+		assertNotNull("SVG template must not be null", svgTemplate);
+		cb.addTemplate(svgTemplate, 50, 500);
+
+		document.newPage();
+		document.close();
+
+		byte[] pdfBytes = pdfOut.toByteArray();
+		assertTrue("PDF must be non-empty", pdfBytes.length > 0);
+		assertTrue("PDF must start with %PDF-",
+				new String(pdfBytes, 0, Math.min(pdfBytes.length, 5), StandardCharsets.ISO_8859_1).startsWith("%PDF-"));
+	}
+
+	public void testSvgEmbeddedAsVectorNotRasterized() throws Exception {
+		byte[] svgData = SVG_WITH_PATH.getBytes(StandardCharsets.UTF_8);
+
+		Document document = new Document(PageSize.A4);
+		ByteArrayOutputStream pdfOut = new ByteArrayOutputStream();
+		PdfWriter writer = PdfWriter.getInstance(document, pdfOut);
+
+		document.open();
+
+		PdfContentByte cb = writer.getDirectContent();
+		PdfTemplate svgTemplate = transSvgToTemplate(svgData, 200, 200);
+		cb.addTemplate(svgTemplate, 50, 500);
+
+		document.close();
+
+		byte[] pdfBytes = pdfOut.toByteArray();
+		String pdfContent = new String(pdfBytes, StandardCharsets.ISO_8859_1);
+
+		assertTrue("PDF must contain vector path commands",
+				pdfContent.contains(" m ") || pdfContent.contains(" l "));
+
+		assertFalse("PDF must NOT contain JPEG raster stream (DCTDecode)",
+				pdfContent.contains("/DCTDecode"));
+	}
+
+	public void testSvgWithNamespaceEmbeddedCorrectly() throws Exception {
+		String svgWithNs = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+				"<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" " +
+				"width=\"100\" height=\"100\">" +
+				"<rect x=\"0\" y=\"0\" width=\"100\" height=\"100\" fill=\"yellow\"/>" +
+				"</svg>";
+		byte[] svgData = svgWithNs.getBytes(StandardCharsets.UTF_8);
+
+		PdfTemplate template = transSvgToTemplate(svgData, 100, 100);
+		assertNotNull("SVG with namespace must transcode to non-null template", template);
+	}
+
+	public void testEmptySvgDataThrowsException() throws Exception {
+		try {
+			transSvgToTemplate(new byte[0], 100, 100);
+			fail("Empty SVG data should cause an exception");
+		} catch (Exception e) {
+		}
+	}
+
+	public void testNullSvgDataDoesNotFail() throws Exception {
+		PdfTemplate template = transSvgToTemplate(null, 100, 100);
+		assertNotNull("Null SVG data should still produce a template (empty)", template);
+	}
+
+	private PdfTemplate transSvgToTemplate(byte[] svgData, float width, float height) throws Exception {
+		ByteArrayOutputStream pdfOut = new ByteArrayOutputStream();
+		Document document = new Document(new org.openpdf.text.Rectangle(width + 20, height + 20));
+		PdfWriter writer = PdfWriter.getInstance(document, pdfOut);
+
+		document.open();
+		PdfContentByte cb = writer.getDirectContent();
+
+		PdfTemplate template = cb.createTemplate(width, height);
+		java.awt.Graphics2D g2D = template.createGraphics(width, height);
+
+		PrintTranscoder transcoder = new PrintTranscoder();
+		if (svgData != null && svgData.length > 0) {
+			transcoder.transcode(new TranscoderInput(new ByteArrayInputStream(svgData)), null);
+		}
+		transcoder.addTranscodingHint(SVGAbstractTranscoder.KEY_ALLOW_EXTERNAL_RESOURCES, Boolean.TRUE);
+
+		java.awt.print.PageFormat pg = new java.awt.print.PageFormat();
+		java.awt.print.Paper p = new java.awt.print.Paper();
+		p.setSize(width, height);
+		p.setImageableArea(0, 0, width, height);
+		pg.setPaper(p);
+		transcoder.print(g2D, pg, 0);
+
+		g2D.dispose();
+		document.close();
+
+		return template;
+	}
+}

--- a/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/util/SvgFileTest.java
+++ b/engine/org.eclipse.birt.report.engine.tests/test/org/eclipse/birt/report/engine/util/SvgFileTest.java
@@ -1,0 +1,159 @@
+/*******************************************************************************
+ * Copyright (c) 2026 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0/.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   See git history
+ *******************************************************************************/
+
+package org.eclipse.birt.report.engine.util;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.nio.charset.StandardCharsets;
+
+import org.junit.Test;
+
+public class SvgFileTest {
+
+	@Test
+	public void testIsSvgByExtension() {
+		assertTrue(SvgFile.isSvg(null, null, ".svg"));
+		assertTrue(SvgFile.isSvg(null, null, ".SVG"));
+		assertTrue(SvgFile.isSvg(null, null, ".SvG"));
+	}
+
+	@Test
+	public void testIsSvgByMimeType() {
+		assertTrue(SvgFile.isSvg("image/svg+xml", null, null));
+		assertTrue(SvgFile.isSvg("IMAGE/SVG+XML", null, null));
+	}
+
+	@Test
+	public void testIsSvgByUriExtension() {
+		assertTrue(SvgFile.isSvg(null, "http://example.com/image.svg", null));
+		assertTrue(SvgFile.isSvg(null, "/path/to/file.SVG", null));
+	}
+
+	@Test
+	public void testIsSvgByUriMimeType() {
+		assertTrue(SvgFile.isSvg(null, "data:image/svg+xml;base64,PHN2Zz4=", null));
+	}
+
+	@Test
+	public void testIsSvgByContent() {
+		String svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"100\" height=\"100\"><rect/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithXmlDeclaration() {
+		String svg = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><svg xmlns=\"http://www.w3.org/2000/svg\"><rect/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithDoctype() {
+		String svg = "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\" \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\"><svg xmlns=\"http://www.w3.org/2000/svg\"><rect/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithXmlAndDoctype() {
+		String svg = "<?xml version=\"1.0\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\" \"\">\n<svg xmlns=\"http://www.w3.org/2000/svg\"><rect/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithXmlAndDoctypeInUtf16() {
+		String svg = """
+				<?xml version="1.0" encoding="UTF-16"?>
+				<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "">
+				<svg xmlns="http://www.w3.org/2000/svg">
+				  <rect/>
+				</svg>
+				""";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_16);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithXmlAndDoctypeAndComment() {
+		String svg = """
+				<?xml version="1.0"?>
+				<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "">
+				<!-- Test comment -->
+				
+				<svg xmlns="http://www.w3.org/2000/svg">
+					<rect/>
+				</svg>
+				""";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentWithLeadingWhitespace() {
+		String svg = "   \n  <svg xmlns=\"http://www.w3.org/2000/svg\"><rect/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, null, svgData));
+	}
+
+	@Test
+	public void testIsSvgByContentCaseInsensitive() {
+		assertTrue(SvgFile.isSvg(null, null, null,
+				"<SVG xmlns=\"http://www.w3.org/2000/svg\"><rect/></SVG>".getBytes(StandardCharsets.UTF_8)));
+		assertTrue(SvgFile.isSvg(null, null, null,
+				"<Svg xmlns=\"http://www.w3.org/2000/svg\"><rect/></Svg>".getBytes(StandardCharsets.UTF_8)));
+	}
+
+	@Test
+	public void testIsSvgByContentNotSvg() {
+		String png = "\211PNG\r\n\032\n";
+		assertFalse(SvgFile.isSvg(null, null, null, png.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	@Test
+	public void testIsNotSvgNonXmlContent() {
+		String jpeg = "\377\330\377";
+		assertFalse(SvgFile.isSvg(null, null, null, jpeg.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	@Test
+	public void testIsNotSvgRandomXml() {
+		String xml = "<?xml version=\"1.0\"?><data><value>test</value></data>";
+		assertFalse(SvgFile.isSvg(null, null, null, xml.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	@Test
+	public void testIsNotSvgNullData() {
+		assertFalse(SvgFile.isSvg(null, null, null, null));
+	}
+
+	@Test
+	public void testIsNotSvgEmptyData() {
+		assertFalse(SvgFile.isSvg(null, null, null, new byte[0]));
+	}
+
+	@Test
+	public void testIsNotSvgTooShortData() {
+		assertFalse(SvgFile.isSvg(null, null, null, "abc".getBytes(StandardCharsets.UTF_8)));
+	}
+
+	@Test
+	public void testIsSvgByContentNotDetectedWhenNoneOfTheOthers() {
+		String svg = "<svg xmlns=\"http://www.w3.org/2000/svg\"><circle cx=\"50\" cy=\"50\" r=\"40\"/></svg>";
+		byte[] svgData = svg.getBytes(StandardCharsets.UTF_8);
+		assertTrue(SvgFile.isSvg(null, null, ".png", svgData));
+	}
+}

--- a/engine/org.eclipse.birt.report.engine/src/org/eclipse/birt/report/engine/util/SvgFile.java
+++ b/engine/org.eclipse.birt.report.engine/src/org/eclipse/birt/report/engine/util/SvgFile.java
@@ -16,14 +16,26 @@ package org.eclipse.birt.report.engine.util;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
+import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
 import java.util.Base64;
 
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.XMLConstants;
+
 import org.apache.batik.transcoder.TranscoderInput;
 import org.apache.batik.transcoder.TranscoderOutput;
 import org.apache.batik.transcoder.image.PNGTranscoder;
+import org.eclipse.birt.core.util.CommonUtil;
+import org.xml.sax.Attributes;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
 
 /**
  * SVG utility class to support SVG images
@@ -39,8 +51,6 @@ public class SvgFile {
 	private static final String URL_PROTOCOL_TYPE_DATA_BASE = ";base64,";
 	private static final String URL_PROTOCOL_TYPE_DATA_UTF8 = ";utf8,";
 
-	static boolean isSvg = false;
-
 	/**
 	 * Validation of SVG image
 	 *
@@ -48,12 +58,7 @@ public class SvgFile {
 	 * @return the validation result of SVG image
 	 */
 	public static boolean isSvg(String uri) {
-		if (uri != null && (uri.endsWith(FILE_EXTENSION_SVG) || uri.toLowerCase().contains(URL_IMAGE_TYPE_SVG))) {
-			isSvg = true;
-		} else {
-			isSvg = false;
-		}
-		return isSvg;
+		return uri != null && (uri.endsWith(FILE_EXTENSION_SVG) || uri.toLowerCase().contains(URL_IMAGE_TYPE_SVG));
 	}
 
 	/**
@@ -65,13 +70,27 @@ public class SvgFile {
 	 * @return the validation result of SVG image
 	 */
 	public static boolean isSvg(String mimeType, String uri, String extension) {
-		isSvg = ((mimeType != null)
+		return isSvg(mimeType, uri, extension, null);
+	}
+
+	/**
+	 * Validation of SVG image.
+	 *
+	 * @param mimeType  mime type to be validated
+	 * @param uri       URI to be validated
+	 * @param extension extension to be validated
+	 * @param data      raw image data (bytes) for content-based detection
+	 * @return the validation result of SVG image
+	 * @since 4.24
+	 */
+	public static boolean isSvg(String mimeType, String uri, String extension, byte[] data) {
+		return ((mimeType != null)
 				&& mimeType.equalsIgnoreCase(URL_IMAGE_TYPE_SVG)) // $NON-NLS-1$
 				|| ((uri != null)
-						&& (uri.toLowerCase().endsWith(FILE_EXTENSION_SVG)
-								|| uri.toLowerCase().contains(URL_IMAGE_TYPE_SVG))) // $NON-NLS-1$
-				|| ((extension != null) && extension.toLowerCase().endsWith(FILE_EXTENSION_SVG)); // $NON-NLS-1$
-		return isSvg;
+				&& (uri.toLowerCase().endsWith(FILE_EXTENSION_SVG)
+				|| uri.toLowerCase().contains(URL_IMAGE_TYPE_SVG))) // $NON-NLS-1$
+				|| ((extension != null) && extension.toLowerCase().endsWith(FILE_EXTENSION_SVG)) // $NON-NLS-1$
+				|| isSvgContent(data);
 	}
 
 	/**
@@ -139,5 +158,85 @@ public class SvgFile {
 		ostream.flush();
 		// use the output stream as Image input stream.
 		return ostream.toByteArray();
+	}
+
+	/**
+	 * Validation of SVG image by content inspection
+	 *
+	 * @param data raw image data (bytes)
+	 * @return true if the content looks like an SVG/XML document
+	 */
+	private static boolean isSvgContent(byte[] data) {
+		if (data == null) {
+			return false;
+		}
+		SvgHandler handler = new SvgHandler();
+		try {
+			SAXParser saxParser = createSecureSaxParser();
+			saxParser.parse(new InputSource(new ByteArrayInputStream(data)), handler);
+			// If we get here, there was no first real XML element.
+			// Then this is not a valid SVG file!
+			return false;
+		} catch (StopParsingException spe) {
+			// Check the handler, whether the first element was the svg-Tag.
+			return handler.svgTagFound();
+		} catch (ParserConfigurationException | SAXException | IOException e) {
+			// all other exceptions a due to broken documents or other problems
+			return false;
+		}
+	}
+
+
+	/**
+	 * The default SAXParser provided by CommonUtil does not allow DOCTYPE declarations at all.
+	 * In order to be able to parse SVGs containing a DOCTYPE declaration, we create our own SAXParser.
+         * We use the recommendations from
+	 *   https://xmltools.github.io/xml-security-essentials
+	 * and
+	 *   https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#jaxp-documentbuilderfactory-saxparserfactory-and-dom4j
+	 * for the case where DTDs are needed.
+	 */
+	private static SAXParser createSecureSaxParser() throws ParserConfigurationException, SAXException {
+		SAXParserFactory spf = SAXParserFactory.newInstance();
+
+		// If DTDs are needed, at minimum disable these:
+		spf.setFeature("http://xml.org/sax/features/external-general-entities", false);  //$NON-NLS-1$
+		spf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);  //$NON-NLS-1$
+		spf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);  //$NON-NLS-1$
+		spf.setXIncludeAware(false);
+		spf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+
+		return spf.newSAXParser();
+	}
+
+	/**
+	 * Custom exception thrown in order to abort further parsing.
+	 */
+	private static class StopParsingException extends SAXException {
+		public StopParsingException(String message) {
+			super(message);
+		}
+	}
+
+	/**
+	 * SAX-Handler to fast determine whether content is a valid SVG file.
+	 */
+	private static class SvgHandler extends DefaultHandler {
+
+		private boolean svgTagFound = false;
+
+		@Override
+		public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
+			if (!svgTagFound && qName.equalsIgnoreCase("svg")) {
+				// this is the "normal" case: first real tag of the document is a svg tag
+				svgTagFound = true;
+			}
+			// End checks fast!
+			throw new StopParsingException("Parsing finished");
+		}
+
+		public boolean svgTagFound() {
+			return this.svgTagFound;
+		}
 	}
 }

